### ISSUE TITLE
Parser is really slow with LuaJIT

### ISSUE THREAD

**Issue by bungle** [vurtun/json#1]
I tried to create bindings to this library with LuaJIT here (decoding json to lua table structures):
https://github.com/bungle/lua-resty-opjson/blob/master/lib/resty/opjson.lua

The parser is quite slow compared to others (maybe it is my code that makes it slow) I refer these microbenchmarks here: https://github.com/bungle/lua-resty-libcjson/issues/3

While other libs I have tried return something between 5.8 - 8.2 secs (os.clock), your json implementation returns something like 30 secs. Making it almost 5 to 8 times slower than others. This is interesting as with say lua-resty-jsmn I have to do multiple passes in parsing.

But one investigation show that all these:

```
struct json_iter json_begin(const json_char*, json_size);
struct json_iter json_read(struct json_token*, const struct json_iter*);
struct json_iter json_parse(json_pair, const struct json_iter*);
```

will throw this warning when executed with `luajit -jv`:

```
NYI: unsupported C function type
```

So it might be that it's just not JIT compilable. Not sure why, yet.

Other than that I would like to have the API changed slightly:

1. return strings unquoted
2. json_token.sub to return immediate childs (not all the descendants) or json_token.childs in addition to that

**Comment by vurtun** [comment]
Ok thanks for your feedback. 
1.) Speed problems:
First did you compile with the -fno-gcse flag? It is a specially designed flag for computed GOTOs and brings better optimization (Enabled at levels -O2, -O3, -Os). If the flag does not solve the problem you could rewrite the json_read function to not return a new json_iter every time and instead update the old one instead, which saves a lot of allocation time in Lua. If this does not solve the problem I will look at the problem and dig deeper into the issue.
2.) Warnings
Interesting I will look into it.
3.) String quotes
Removing the String quotes is quite easy but I use the string quotes in the json_type function to better distinguish numbers and strings so it would complicate the type check immensely. But a easier solution would be to just increment the json_token.str and decrement the json_token.len attribute. Again if the proposed solution is not good enough I will have to dig deeper and find a better way to distinguish types from each other. 
4.) Children
Good idea, should no be that hard to implement.

**Comment by vurtun** [comment]
Small update, I added the number of children to struct json_token and provided a way to remove the quotes from string tokens. As for the speed problems and the LuaJIT warnings I will look deeper into the issue either tomorrow or on the weekend.

**Comment by bungle** [comment]
Hi,

this is how I compiled the library (on OSX, clang gives about the same results and some compilation warnings):

```
/usr/local/bin/gcc-4.9 json.c -o libopjson.so -shared -fno-gcse -fno-crossjumping -O3
```

I also added support for json_deq and support to allocate right sized tables with LuaJIT 2.1 (table.new) using token.children.

Some comments:

1. I now get in my benchmark something like 26.55 secs (still rather slow).
2. I still get NYI warnings
3. There is no way to get number of children in a root object?

I still have not tried to rewrite `json_read` function to reuse the iterator (but I'm quite sure it doesn't help here).

This jsmn [1] library is also horribly slow if not compiled with `-DJSMN_PARENT_LINKS`, but when it is compiled with that, it is pretty fast (and seems to work better with LuaJIT).

[1] http://zserge.com/jsmn.html

**Comment by vurtun** [comment]
Hey,
ok so now I know the speed problem is on my side of the equation or the lua code. I will
definitely take a deeper look at where the bottleneck is. There are a few possible problem
areas that are rather hard to detect like the possibility that I destroy the cache with my computed addresses or the amount of reparsing. As for the nice compile flag in jsmn I personally don't use hierarchies so there is no hack like that for me. 

On the Lua side did you run the LuaJIT profiler that shipped with version 2.1 or any other profiler? It would be nice to know if there is one bottleneck or if there is a bigger issue with my implementation.

I will definitely test the library and the lua code tomorrow. Do you have a special JSON file you do the testing on?

**Comment by bungle** [comment]
The benchmark code I'm running is like this:

```lua
local open, clock = io.open, os.clock
local d = require "resty.opjson".decode

local file = open("citylots.json", "rb")
local content = file:read "*a"
file:close()

local x, z, l = 0, 0, #content
x = clock()
d(content)
z = clock() - x
print(string.format("D1 Decoding Time: %.6f", z))
```

The `citylots.json` can be found from here:
https://raw.githubusercontent.com/zemirco/sf-city-lots-json/master/citylots.json

It is a really big one. I also tried smaller json-file with manu iterations but I got similar results.

I have run the benchmarks with LuaJIT 2.0.3, latest LuaJIT 2.1 alpha, and OpenResty resty (command), which is the latest LuaJIT 2.1 alpha with some OpenResty specific stuff in it (somehow the resty binary seems to be the fastest, but it might have something to do with LuaJIT compile options or something).

The NYI messages (meaning LuaJIT has not yet implemented JIT on C-functions that your library exports - I'm not sure what the limitations exactly are). But then again, I can get quite good speed when JIT is turned off with others (I have to recheck these though). The problem might be in my Lua code as well (it uses recursive function calls - but I have used similar techniques with others as well).

**Comment by vurtun** [comment]
Hey
I am currently writing a test parser in C for the citylots.json file and getting around 1.5 seconds
by doing a iterative pass over the JSON file. I am not completely done yet (reading the coordinates gave me a strange error in my benchmark code) but I think it already shows that the pass is far away from the 26 seconds. Of course Lua is a bit slower than C but I don't think to that extent. I will try to complete the benchmark but I heavily suspect that the problem lies on the Lua side.

**Comment by bungle** [comment]
Hi, did you get 3.5 secs with my Lua code in you benchmark (it seems about correct - what could cause my machine to slow down)? Could there be something specific on OSX then (I'm using 10.9.5)?

Other than that I found a bug in the children count code. With this json:

```javascript
{
    "b": { "a": {}, "b": [], "c": "test", "d": true, "e": false,  "f": null, "g": 10 },
    "a": [ {}, [], "test", true, false, null, 10 ]
}
```

-> children count of "a" is 5, should be 7 (arrays, and objects are not counted).
-> children count of "b" is 12 (values are counted as well, but array values and object values are not), should be 7 (the number of keys).

**Comment by vurtun** [comment]
Hey,
1.) Benchmark
I achieved the 3.5 seconds with a C benchmark on my linux machine NOT with your Lua benchmark. I wanted to make sure that the performance problem was not in the library. I will try out your Lua benchmark now and see what I can find. Do you know if it is prohibited to return a complete stack object in a function with LuaJIT like I do with json_read, json_parse and json_begin? Maybe that is the origin of the LuaJIT warnings.

2.) Children
Fixed the children count. I previously only counted the tokens not the actual pairs and values and even that had a small bug, but should be fixed now.

3.) Number of children in root node
I would have to parse the complete file just to get the number of children. So while it is possible in theory it would cost me around 400 ms in the citylots.json file to get them. So I don't know if the gain of knowing the number of children in the root outweighs the cost.

**Comment by bungle** [comment]
Now I hit yet another bug:

This json:

```javascript
{
    "type": "FeatureCollection",
    "features": [
        {
            "type": "Feature",
            "properties": {
                "MAPBLKLOT": "0001001",
                "BLKLOT": "0001001",
                "BLOCK_NUM": "0001",
                "LOT_NUM": "001",
                "FROM_ST": "0",
                "TO_ST": "0",
                "STREET": "UNKNOWN",
                "ST_TYPE": null,
                "ODD_EVEN": "E"
            },
            "geometry": {
                "type": "Polygon",
                "coordinates": [
                    [
                        [ -122.422003528252475, 37.808480096967251, 0.0 ],
                        [ -122.422076013325281, 37.808835019815085, 0.0 ],
                        [ -122.421102174348633, 37.808803534992904, 0.0 ],
                        [ -122.421062569067274, 37.808601056818148, 0.0 ],
                        [ -122.422003528252475, 37.808480096967251, 0.0 ]
                    ]
                ]
            }
        }
    ]
}
```

I doesn't return correct `json_token.str`s.

**Comment by bungle** [comment]
Hi, I will check my code. There seems to be some problems in it.

**Comment by vurtun** [comment]
Small update. I tried your Lua code and got about 9 - 10 seconds with Arch Linux x86_64, Intel Core i5-4210U @1.70GHz, 4GB DD3 RAM and SSD. I got the same warning as you for the three functions.

**Comment by bungle** [comment]
Yes, but that 10 secs result is from the latest opjson.lua (from github) that doesn't decode correct results. I found that there is a problem in your code for children counts, e.g. try this:

```javascript
{
    "type": "FeatureCollection",
    "features": [
        {
            "type": "Feature",
            "properties": {
                "MAPBLKLOT": "0001001",
                "BLKLOT": "0001001",
                "BLOCK_NUM": "0001",
                "LOT_NUM": "001",
                "FROM_ST": "0",
                "TO_ST": "0",
                "STREET": "UNKNOWN",
                "ST_TYPE": null,
                "ODD_EVEN": "E"
            },
            "geometry": {
                "type": "Polygon",
                "coordainates": [
                    [
                        [ -122.422003528252475, 37.808480096967251, 0.0 ],
                        [ -122.422076013325281, 37.808835019815085, 0.0 ],
                        [ -122.421102174348633, 37.808803534992904, 0.0 ],
                        [ -122.421062569067274, 37.808601056818148, 0.0 ],
                        [ -122.422003528252475, 37.808480096967251, 0.0 ]
                    ]
                ]
            }
        }
    ]
}
```

At first it looks ok, but then it starts giving 0 for some sizes.

**Comment by vurtun** [comment]
Yeah I fucked up the conditional check if the token has the right depth, should be fixed now.

**Comment by bungle** [comment]
It looks that now `json_typ` returns wrong types.

**Comment by vurtun** [comment]
Hm, I can't find any problem. Did you do json_deq before you type check because that
destroys the ability to check the type. Or if that is not the problem could you give an example where and when this behavior happens.

**Comment by bungle** [comment]
I will do further investigation.

**Comment by bungle** [comment]
Simple example:

JSON:

```javascript
{
    "test": 13
}
```

Test code:

```lua
local json = [[
{
    "test": 13
}
]]
local p = ffi_new("json_pair")
local i = lib.json_begin(json, #json)
lib.json_parse(p, i)
print(tonumber(lib.json_type(p[0])), tonumber(lib.json_type(p[1])))
```

This will print out:

```
4    0
```

It should print:

```
4    3
```

This was alright before your latest fix to children counts. I also checked that the p[0] is "test", and it was.

**Comment by vurtun** [comment]
That is absolutely strange. I have the exact same test in the C unit test and for me everything
works fine and gives the right result. But if I try your Lua program the wrong result gets returned.

**Comment by bungle** [comment]
Yes it is, at least because it was working earlier. I updated my code, but the results are currently:

```
Decoding Time: 31.641169
```

**Comment by vurtun** [comment]
At the moment I only get out of memory errors, so I can't even finish your current or the
older Lua benchmark. The C benchmark on the other hand got faster with around 3.2 seconds.
I just don't have enough experience with Lua and LuaJIT so I have currently no idea what the problem could be.

**Comment by bungle** [comment]
Hi, I have now rewrote the parser to reuse iterator, now I get something like:

```
Decoding Time: 8.900796
```

This is a fork. I will publish the changes as soon as I get to finnish them. Looks good so far (now it is almost inline with others.

**Comment by bungle** [comment]
I'm happy to tell you that now I got rid of all the NYIs, and the results is now:

```
Decoding Time: 6.001310
```

Which is really great! I will look at optimizing this even more. Let's see if I can get it even faster with LuaJIT.

**Comment by vurtun** [comment]
Whoa amazing. All the hard work has paid of. Interesting to know that LuaJIT does not like stack objects I did not know that.
The time looks really good is seems to be on par with most of the other json libraries.

**Comment by bungle** [comment]
This is what I came up with:
https://github.com/bungle/libopjson (looks like I only need json_num and json_read).

And the new lua-resty-opjson is also checked in:
https://github.com/bungle/lua-resty-opjson/blob/master/lib/resty/opjson.lua

I ended up with:

```
Decoding Time: 5.820199
```

Now I think I just need a more clever way to iterate on Lua side and do away from recursion.

**Comment by vurtun** [comment]
Interesting you seemed to have found your way through my code pretty well, looks good.
Is the time 5.82 seconds with LuaJIT or just with the interpreter and what time was the fastest
for any json parser?

**Comment by bungle** [comment]
The 5.82 secs result is with LuaJIT 2.1-alpha. Normal Lua does not have FFI (built-in), but it has an C-API. C-API cannot be JIT compiled with LuaJIT. From my testing this is now one of the fastest, but guys at Cloudflare have been doing impressive stuff since I started testing these, and because of that, :-):
https://github.com/cloudflare/lua-resty-json/issues/1 (they have not yet checked in the code where those improvements are). But right now your code with the modifications is one of the fastest when integrated with LuaJIT (at least out of those that I have tested or tried implementing).
